csrc/attention/paged_attention_v1.cu

```diff
@@ -773,22 +769,32 @@ void paged_attention_v1_launcher(
   }
 }
 
-#define CALL_V1_LAUNCHER(T, CACHE_T, BLOCK_SIZE, KV_DTYPE, IS_BLOCK_SPARSE)  \
-  paged_attention_v1_launcher<T, CACHE_T, BLOCK_SIZE, KV_DTYPE,              \
-                              IS_BLOCK_SPARSE>(                              \
-      out, query, key_cache, value_cache, num_kv_heads, scale, block_tables, \
-      seq_lens, max_seq_len, alibi_slopes, k_scale, v_scale, tp_rank,        \
-      blocksparse_local_blocks, blocksparse_vert_stride,                     \
+#define CALL_V1_LAUNCHER(T, CACHE_T, BLOCK_SIZE, KV_DTYPE, IS_BLOCK_SPARSE, NUM_THREADS)  \
+  paged_attention_v1_launcher<T, CACHE_T, BLOCK_SIZE, KV_DTYPE,                           \
+                              IS_BLOCK_SPARSE, NUM_THREADS>(                              \
+      out, query, key_cache, value_cache, num_kv_heads, scale, block_tables,              \
+      seq_lens, max_seq_len, alibi_slopes, k_scale, v_scale, tp_rank,                     \
+      blocksparse_local_blocks, blocksparse_vert_stride,                                  \
       blocksparse_block_size, blocksparse_head_sliding_step);
 
-#define CALL_V1_LAUNCHER_SPARSITY(T, CACHE_T, BLOCK_SIZE, IS_FP8_KV_CACHE) \
-  switch (is_block_sparse) {                                               \
-    case true:                                                             \
-      CALL_V1_LAUNCHER(T, CACHE_T, BLOCK_SIZE, IS_FP8_KV_CACHE, true);     \
-      break;                                                               \
-    case false:                                                            \
-      CALL_V1_LAUNCHER(T, CACHE_T, BLOCK_SIZE, IS_FP8_KV_CACHE, false);    \
-      break;                                                               \
+#define CALL_V1_LAUNCHER_W_NUM_THREADS(T, CACHE_T, BLOCK_SIZE, IS_FP8_KV_CACHE, IS_BLOCK_SPARSE) \
+  switch (num_threads) {                                                                  \
+    case 128:                                                                             \
+      CALL_V1_LAUNCHER(T, CACHE_T, BLOCK_SIZE, IS_FP8_KV_CACHE, IS_BLOCK_SPARSE, 128);    \
+      break;                                                                              \
+    case 1024:                                                                            \
+      CALL_V1_LAUNCHER(T, CACHE_T, BLOCK_SIZE, IS_FP8_KV_CACHE, IS_BLOCK_SPARSE, 1024);   \
+      break;                                                                              \
+  }
+
+#define CALL_V1_LAUNCHER_SPARSITY(T, CACHE_T, BLOCK_SIZE, IS_FP8_KV_CACHE)            \
+  switch (is_block_sparse) {                                                          \
```

Comment:
Is it better we add 'is_block_sparse' as an argument?

Reply:
don't think so, following style that file has.

Reply:
A macro using variables not listed in argument list is error-prone and could cause unexpected problems. How about add a new 'IS_BLOCK_SPARSE' to the argument list?

Reply:
this is the whole intent of construction here. Macro is a text substitution before parsing starts, thus is_block_sparse is taken from the context where CALL_V1_LAUNCHER_SPARSITY is used.

Reply:
I know how macro works, so I think this is not the best practice because it is hard to understand and maintain. For example here, contributors not familiar with this macro (like me, haha) may not be able to notice that is_block_sparse in the same scope will affect the output of this macro, so it will be hard for them to understand the real logic, unless they find the definition of the macro. What do you think?

Reply:
That's a design choice by the original vllm, which I'm also not a fan of, but that's how things are. The best approach to changing this would be through proposing a change in the upstream repo.

Reply:
OK then. Thanks @gshtras for providing the context.